src/WebJobs.Script/Binding/ExtensionLoader.cs

```diff
@@ -0,0 +1,224 @@
+                // By convention, restrict to based on filenames.
+                var filename = Path.GetFileName(path);
+                if (!filename.ToLowerInvariant().Contains("extension"))
+                {
+                    continue;
+                }
+
+                try
+                {
+                    // See GetNugetPackagesPath() for details
+                    // Script runtime is already setup with assembly resolution hooks, so use LoadFrom
+                    Assembly assembly = Assembly.LoadFrom(path);
+                    LoadExtensions(assembly, path);
+                }
+                catch (Exception e)
+                {
+                    string msg = $"Failed to load custom extension from '{path}'.";
+                    _traceWriter.Error(msg, e);
+                    _startupLogger.LogError(0, e, msg);
+                }
+            }
+        }
+
+        // Load extensions that are directly referenced by the user types.
+        public void LoadDirectlyReferencedExtensions(IEnumerable<Type> userTypes)
+        {
+            var possibleExtensionAssemblies = UserTypeScanner.GetPossibleExtensionAssemblies(userTypes);
+
+            foreach (var kv in possibleExtensionAssemblies)
+            {
+                var assembly = kv.Key;
+                var locationHint = kv.Value;
+                LoadExtensions(assembly, locationHint);
+            }
+        }
+
+        private void LoadExtensions(Assembly assembly, string locationHint)
+        {
+            foreach (var type in assembly.ExportedTypes)
+            {
+                if (!typeof(IExtensionConfigProvider).IsAssignableFrom(type))
+                {
+                    continue;
+                }
+
+                if (IsExtensionLoaded(type))
+                {
+                    continue;
+                }
+
+                var extensionConfigProvider = (IExtensionConfigProvider)Activator.CreateInstance(type);
+                LoadExtension(extensionConfigProvider, locationHint);
+            }
+        }
+
+        private bool IsExtensionLoaded(Type type)
+        {
+            var registry = _config.HostConfig.GetService<IExtensionRegistry>();
+            var extensions = registry.GetExtensions<IExtensionConfigProvider>();
+            foreach (var extension in extensions)
+            {
+                var loadedExtentionType = extension.GetType();
+                if (loadedExtentionType == type)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private void LoadExtension(IExtensionConfigProvider extensionConfigProvider, string locationHint = null)
+        {
+            var extensionConfigProviderType = extensionConfigProvider.GetType();
+            string extensionName = extensionConfigProviderType.Name;
+
+            string msg = null;
+            if (!string.IsNullOrEmpty(locationHint))
```

Comment:
Improving the log message. Previously we were getting logs like:

```
```
Loaded custom extension: BotFrameworkConfiguration from ''
Loaded custom extension: SendGridConfiguration from ''
Loaded custom extension: EventGridExtensionConfig from ''
```
```